WMF Framework/WMFArticle+Migration.swift

```diff
@@ -0,0 +1,392 @@
+
+        guard let pages: [CDPage] = try wmfDataStore.fetch(
+            entityType: CDPage.self,
+            predicate: predicate,
+            fetchLimit: 1,
+            in: wmfContext
+        ), let page = pages.first else {
+            DDLogError("[SavedPagesMigration] Unsave: no CDPage for PID \(pageIDs.projectID) nsID \(pageIDs.namespaceID) title \(pageIDs.title)")
+            return
+        }
+
+        if let saved = page.savedInfo {
+            wmfContext.delete(saved)
+            page.savedInfo = nil
+        } else {
+            DDLogInfo("[SavedPagesMigration] Unsave no-op: savedInfo already nil for title \(pageIDs.title)")
+        }
+
+        if page.timestamp == nil {
+            DDLogInfo("[SavedPagesMigration] Deleting CDPage for \(pageIDs.title) — timestamp nil after unsave")
+        }
+    }
+
+    private static func applySavedStateOnWMFContext(snapshot: SavedArticleSnapshot, in wmfContext: NSManagedObjectContext, wmfDataStore: WMFCoreDataStore) throws {
+        let predicate = makePredicate(for: snapshot.ids)
+
+        guard let page = try wmfDataStore.fetchOrCreate(entityType: CDPage.self,
+                                                        predicate: predicate,
+                                                        in: wmfContext) else {
+            DDLogError("[SavedPagesMigration] save: no CDPage for PID \(snapshot.ids.projectID) nsID \(snapshot.ids.namespaceID) title \(snapshot.ids.title)")
+            return
+        }
+
+        page.title = snapshot.ids.title
+        page.namespaceID = snapshot.ids.namespaceID
+        page.projectID = snapshot.ids.projectID
+        page.timestamp = snapshot.viewedDate ?? snapshot.savedDate
+
+        if let existing = page.savedInfo {
+            existing.savedDate = snapshot.savedDate
+        } else {
+            let saved = CDPageSavedInfo(context: wmfContext)
+            saved.savedDate = snapshot.savedDate
+            page.savedInfo = saved
+        }
+    }
+
+    private static func makePredicate(for ids: PageIDs) -> NSPredicate {
+        NSPredicate(format: "projectID == %@ AND namespaceID == %d AND title == %@",
+                    ids.projectID, ids.namespaceID, ids.title)
+    }
+
+    // MARK: - Delete all
+
+    private func clearAllSavedData() {
+        guard let wmfDataStore = WMFDataEnvironment.current.coreDataStore else {
+            DDLogError("[SavedPagesMigration] Missing WMFData store")
+            return
+        }
+
+        guard let wmfContext = try? wmfDataStore.newBackgroundContext else {
+            DDLogError("[SavedPagesMigration] Could not create WMFData background context")
+            return
+        }
+        wmfContext.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+
+        wmfContext.perform {
+            do {
+                let savedInfoFR = NSFetchRequest<NSFetchRequestResult>(entityName: "CDPageSavedInfo")
+                let deleteSavedInfo = NSBatchDeleteRequest(fetchRequest: savedInfoFR)
```

Comment:
I am reading around that there may be a chance here that the CDPage.savedInfo relationship may still point to the old CDPageSavedInfo deleted entity, so it'll be invalid. Can you null the relationship out if that's the case? You might be able to use a `NSBatchUpdateRequest` to update all savedInfo properties to null.